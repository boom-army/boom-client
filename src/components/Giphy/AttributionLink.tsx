import React, { useContext } from "react";
import Link from "@mui/material/Link";
import { styled } from "@mui/system";
import { ThemeContext } from "../../contexts/theme";
import { darkTheme } from "../../styles/themes";

export const AttributionLink: React.FC<{ src: any |null}> = ({ src }) => {
  const { theme } = useContext(ThemeContext);
  const isDarkTheme = theme.primaryColor === darkTheme.primaryColor;

  const GiphyAttributionImg = styled("img")({
    boxSizing: "content-box",
    borderBottom: "2px solid transparent",
    paddingBottom: "3px",
    "&:hover": {
      borderBottom: `2px solid ${theme.tertiaryColor}`,
    },
  });

  return (
    <Link href={src} target="_blank">
      <GiphyAttributionImg
        width="150px"
        src={`/assets/giphy-${isDarkTheme ? "dark" : "light"}.png`}
        alt="Powered by Giphy"
      />
    </Link>
  );
};
