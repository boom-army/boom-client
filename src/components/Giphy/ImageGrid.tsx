import React from "react";
import ImageList from "@mui/material/ImageList";
import ImageListItem from "@mui/material/ImageListItem";
import { Box } from "@mui/system";
import { GIFObject } from "giphy-api";
import { Loader } from "../Loader";
import Stack from "@mui/material/Stack";

interface Props {
  gifArr: GIFObject[];
  setGif: React.Dispatch<React.SetStateAction<GIFObject>>;
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
  isLoadingMore: boolean;
}

export const ImageGrid: React.FC<Props> = ({
  gifArr,
  setGif,
  setOpen,
  isLoadingMore,
}) => {
  if (!gifArr?.length) {
    return (
      <Box sx={{ margin: 2 }}>
        <p>No gifs were found</p>
      </Box>
    );
  }
  return (
    <Stack direction="column">
      <ImageList variant="masonry" cols={3} gap={8} sx={{ overflowY: "auto" }}>
        {gifArr.map((item) => (
          <ImageListItem
            component="button"
            key={item.id}
            onClick={() => {
              setGif(item);
              setOpen(false);
            }}
            sx={{
              margin: 0,
              padding: 0,
              appearance: "none",
              background: "none",
              border: "none",
              lineHeight: 0,
            }}
          >
            <img
              src={item.images.fixed_width_downsampled.url}
              alt={item.title}
            />
          </ImageListItem>
        ))}
      </ImageList>
      <Box height="50px">{isLoadingMore && <Loader />}</Box>
    </Stack>
  );
};
