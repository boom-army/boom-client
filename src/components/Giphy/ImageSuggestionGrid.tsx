import React from "react";
import ImageList from "@mui/material/ImageList";
import ImageListItem from "@mui/material/ImageListItem";
import ImageListItemBar from "@mui/material/ImageListItemBar";
import { GIPHY_SEARCH_SUGGESTIONS } from "../../constants/giphy-search-suggestions";

type SetInput = React.Dispatch<React.SetStateAction<string>>;
type SetIsLoading = React.Dispatch<React.SetStateAction<boolean>>;

export const ImageSuggestionGrid: React.FC<{
  setInput: SetInput;
  setIsLoading: SetIsLoading;
}> = ({ setInput, setIsLoading }) => {
  return (
    <>
      <ImageList cols={2} gap={2} sx={{ overflowY: "hidden" }}>
        {GIPHY_SEARCH_SUGGESTIONS.map((item, index) => {
          // If the last item is odd, then make it fill the entire row
          const isLastItem = index === GIPHY_SEARCH_SUGGESTIONS.length - 1;
          const isLastItemOdd = GIPHY_SEARCH_SUGGESTIONS.length % 2 === 1;
          const cols = isLastItem && isLastItemOdd ? 2 : 1;
          const rows = isLastItem && isLastItemOdd ? 2 : 1;

          return (
            <ImageListItem
              key={item.id}
              cols={cols}
              rows={rows}
              sx={{
                height: "150px!important",
                overflow: "hidden",
                cursor: "pointer",
              }}
              onClick={() => {
                setIsLoading(true);
                setInput(item.title);
              }}
            >
              <img src={item.image} alt={item.title} loading="lazy" />
              <ImageListItemBar
                sx={{
                  background:
                    "linear-gradient(to bottom, rgba(0,0,0,0.7) 0%, " +
                    "rgba(0,0,0,0.3) 70%, rgba(0,0,0,0) 100%)",
                }}
                title={item.title}
                position="top"
                actionPosition="left"
              />
            </ImageListItem>
          );
        })}
      </ImageList>
    </>
  );
};
