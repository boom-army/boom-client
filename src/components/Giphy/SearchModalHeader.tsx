import React, { useContext } from "react";
import CloseIcon from "@mui/icons-material/Close";
import IconButton from "@mui/material/IconButton";
import { FormControl, InputAdornment, useTheme } from "@mui/material";
import { TextField } from "@mui/material";

interface Props {
  input: string;
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setInput: React.Dispatch<React.SetStateAction<string>>;
  handleClose: () => void;
}

export const SearchModalHeader: React.FC<Props> = ({
  input,
  setIsLoading,
  setInput,
  handleClose,
}) => {
  const theme = useTheme();
  return (
    <FormControl fullWidth={true}>
      <TextField
        hiddenLabel={true}
        fullWidth={true}
        type="text"
        autoFocus
        placeholder="Search for GIF"
        value={input}
        onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
          setIsLoading(true);
          setInput(e.target.value);
        }}
        InputProps={{
          style: { color: theme.palette.primary, padding: 10 },
          endAdornment: (
            <InputAdornment position="end">
              <IconButton
                onClick={handleClose}
                aria-label="close"
                size="medium"
                disableRipple={true}
              >
                <CloseIcon sx={{ color: theme.accentColor }} />
              </IconButton>
            </InputAdornment>
          ),
        }}
      />
    </FormControl>
  );
};
