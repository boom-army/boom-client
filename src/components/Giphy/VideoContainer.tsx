import React, { useRef } from "react";
import { styled } from "@mui/material/styles";
import { Box } from "@mui/system";
import IconButton from "@mui/material/IconButton";
import CloseIcon from "@mui/icons-material/Close";

interface Gif {
  id?: string;
  fixedHeightUrl?: string;
  originalUrl?: string;
  title?: string;
  url?: string;
}

interface VideoProps {
  gif: Gif;
  onClose?: (() => void) | undefined;
}

const VideoBoxWrapper = styled("div")(({ theme }) => ({
  display: "flex",
  flexDirection: "column",
  alignItems: "center",
  justifyContent: "center",
  position: "relative",
  borderRadius: "16px",
  border: `1px solid ${theme.palette.secondary}`,
  width: "90%",
  margin: "0 0 0.75rem",
  overflow: "hidden",
}));

const Video = styled("video")({
  width: "100%",
  cursor: "auto",
});

export const VideoContainer = React.memo(({ gif, onClose }: VideoProps) => {
  const { fixedHeightUrl, title } = gif;
  const videoRef = useRef<HTMLVideoElement>(null);

  const handleVideoPlayerClick = () => {
    const video = videoRef.current;
    if (video && video.paused) {
      video.play().catch((error) => {
        console.error("Video playback failed:", error);
      });
    }
  };

  return (
    <VideoBoxWrapper>
      {onClose && (
        <Box
          sx={{
            position: "absolute",
            left: "3%",
            top: "3%",
            padding: "5px",
            borderRadius: "50%",
            background: "rgba(0, 0, 0, 0.7)",
            lineHeight: "1.5",
            fontSize: "14px",
            zIndex: "99",
            "&:hover": {
              background: "rgba(0, 0, 0, 0.57)",
            },
          }}
        >
          <IconButton onClick={onClose} aria-label="close" size="medium">
            <CloseIcon />
          </IconButton>
        </Box>
      )}

      <Video
        ref={videoRef}
        onClick={handleVideoPlayerClick}
        playsInline
        autoPlay
        muted
        loop
        aria-label={title}
      >
        <source src={fixedHeightUrl} type="video/mp4" />
      </Video>

      <Box
        sx={{
          position: "absolute",
          left: "3%",
          bottom: "3%",
          padding: "0 4px",
          borderRadius: "4px",
          background: "rgba(0, 0, 0, 0.77)",
          lineHeight: "1.5",
          fontSize: "14px",
        }}
      >
        GIF
      </Box>
    </VideoBoxWrapper>
  );
});
