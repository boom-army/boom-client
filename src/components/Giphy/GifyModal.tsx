import React, {
  useState,
  useEffect,
  useContext,
  useCallback,
  useRef,
} from "react";
import { Box } from "@mui/system";
import { GifIcon } from "../Icons";
import { GiphyContext, Search } from "../../contexts/giphy";
import { ImageGrid } from "./ImageGrid";
import { ImageSuggestionGrid } from "./ImageSuggestionGrid";
import { Loader } from "../Loader";
import { SearchModalHeader } from "./SearchModalHeader";
import { useTheme } from '@mui/material/styles';
import { debounce } from "lodash";
import { IconButton, Stack, Modal } from "@mui/material";
import { blue, grey } from "@mui/material/colors";

const GIPHY_API = process.env.REACT_APP_GIPHY_KEY as string;

const queryGiphy = async (
  query: string,
  offset: number = 0
): Promise<Search> => {
  const queryParams = new URLSearchParams({
    api_key: GIPHY_API,
    offset: offset.toString(),
    q: query,
  });
  const response = await fetch(
    `https://api.giphy.com/v1/gifs/search?${queryParams}`
  );
  const json = await response.json();
  return { query, gif: json.data };
};

export const GifyModal: React.FC<{
  setGif: any;
}> = ({ setGif }) => {
  const [input, setInput] = useState("");
  const [open, setOpen] = useState(false);
  const [isLoading, setIsLoading] = useState(false);
  const [isLoadingMore, setIsLoadingMore] = useState(false);
  const [error, setError] = useState<any>(null);
  const [searchResults, setSearchResults] = useState<Search | null>(null); // current searched gifs
  const { searchGiphyCache, setSearchGiphyCache } = useContext(GiphyContext); // search cache
  const theme = useTheme();
  const handleClose = () => setOpen(false);
  const handleOpen = () => {
    setInput("");
    setError(false);
    setSearchResults(null);
    setOpen(true);
  };

  const imageBoxRef = useRef<any>(null);

  const handleScroll = debounce(async () => {
    if (!imageBoxRef.current || !input) return;
    const { scrollHeight, scrollTop, clientHeight } = imageBoxRef.current;

    if (scrollHeight - scrollTop - clientHeight >= 1) return;

    try {
      const currentSearchResultsLength = searchResults?.gif.length;
      // the initial search returned no hits so there is no point querying again
      if (currentSearchResultsLength === 0) return;

      setIsLoadingMore(true);

      const query = input.toLowerCase();
      const result = await queryGiphy(query, currentSearchResultsLength);

      if (!result?.gif.length) return;

      const updatedSearchCache = searchGiphyCache.map((cache) => {
        if (cache.query === result.query) {
          return {
            query: cache.query,
            gif: [...cache.gif, ...result.gif],
          };
        }
        return cache;
      });

      setSearchGiphyCache(updatedSearchCache);
      const updatedGifs =
        updatedSearchCache.find((cache) => cache.query === query) || null;
      setSearchResults(updatedGifs);
    } finally {
      setIsLoadingMore(false);
    }
  }, 300);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const memoDebouncedScrollHandler = useCallback(handleScroll, [
    searchResults,
    searchGiphyCache,
    input,
  ]);

  useEffect(() => {
    const imageListNode = imageBoxRef.current;

    if (!imageListNode) return;

    imageListNode.addEventListener("scroll", memoDebouncedScrollHandler);
    return () => {
      imageListNode.removeEventListener("scroll", memoDebouncedScrollHandler);
    };
  }, [memoDebouncedScrollHandler]);

  const handleInputQuery = debounce(
    async (value: string, searchGiphyCache: Search[]) => {
      setError(false);

      const query = value.toLowerCase();

      if (!query) {
        setIsLoading(false);
        return setSearchResults(null);
      }

      const cachedGif = searchGiphyCache.find((gif) => gif.query === query);

      if (cachedGif) {
        setIsLoading(false);
        return setSearchResults(cachedGif);
      }

      try {
        const searchResult = await queryGiphy(query);
        setSearchGiphyCache([...searchGiphyCache, searchResult]);
        setSearchResults(searchResult);
        setError(false);
      } catch (error) {
        setError(true);
      } finally {
        setIsLoading(false);
      }
    },
    800
  );

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const memoDebouncedQuery = useCallback(handleInputQuery, []);

  useEffect(() => {
    memoDebouncedQuery(input, searchGiphyCache);
  }, [input, memoDebouncedQuery, searchGiphyCache]);

  return (
    <>
      <IconButton onClick={handleOpen}>
        <GifIcon />
      </IconButton>
      <Modal
        open={open}
        onClose={handleClose}
      >
        <Box
          sx={{
            position: "absolute",
            top: "30%",
            transform: "translate(20%, -60%)",
            bgcolor: theme.palette.background.default,
            border: `2px solid ${blue[100]}`,
            borderRadius: 1,
            maxWidth: "90vw",
            width: 500,
            height: 650,
            minHeight: 500,
            maxHeight: "90vh",
          }}
        >
          <Stack direction="column" sx={{ height: "100%" }}>
            <SearchModalHeader
              input={input}
              setIsLoading={setIsLoading}
              setInput={setInput}
              handleClose={handleClose}
            />
            <Box
              ref={imageBoxRef}
              sx={{ height: "100%", overflowY: "auto", overflowX: "hidden" }}
            >
              {isLoading || error ? (
                <Box sx={{ margin: 2, position: "relative" }}>
                  {isLoading ? (
                    <Loader />
                  ) : (
                    <p>There was an error while searching.</p>
                  )}
                </Box>
              ) : (
                <>
                  {searchResults && (
                    <ImageGrid
                      gifArr={searchResults.gif}
                      setGif={setGif}
                      setOpen={setOpen}
                      isLoadingMore={isLoadingMore}
                    />
                  )}
                  {!searchResults && (
                    <ImageSuggestionGrid
                      setInput={setInput}
                      setIsLoading={setIsLoading}
                    />
                  )}
                </>
              )}
            </Box>
          </Stack>
        </Box>
      </Modal>
    </>
  );
};
