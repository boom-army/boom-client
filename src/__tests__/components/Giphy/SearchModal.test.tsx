import React from "react";
import { SearchModal } from "../../../components/Giphy/SearchModal";
import { act, fireEvent, render, screen } from "@testing-library/react";
import { createSerializer } from '@emotion/jest'

expect.addSnapshotSerializer(createSerializer(
  {
    classNameReplacer(className, index) {
      return `css-${index}`
    },
        // includeStyles: false 
  }
));

let setGif: any = "";

describe("<SearchModal/> component :", () => {
  test("display <SearchModal/> component  when Search result found ", async () => {
    let rendered = render(<SearchModal setGif={setGif} />);
    expect(rendered).toMatchSnapshot();
  });
});
test("Assertion testing of <SearchModal/>  component when Search result found", async () => {
  render(<SearchModal setGif={setGif} />);
  expect(screen.getByRole("button")).toBeInTheDocument();
});
