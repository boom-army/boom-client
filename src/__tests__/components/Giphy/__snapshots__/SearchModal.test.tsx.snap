// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchModal/> component : display <SearchModal/> component  when Search result found  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .css-0 {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
}

<body>
    <div>
      <button
        class="tweet-gif css-0"
        type="button"
      >
        <svg
          fill="#C4C4C4"
          height="24"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M3 2H21C21.5523 2 22 2.44772 22 3V21C22 21.5523 21.5523 22 21 22H3C2.44772 22 2 21.5523 2 21V3C2 2.44772 2.44772 2 3 2ZM0 3C0 1.34315 1.34315 0 3 0H21C22.6569 0 24 1.34315 24 3V21C24 22.6569 22.6569 24 21 24H3C1.34315 24 0 22.6569 0 21V3ZM9.61279 15.8228C10.1141 15.633 10.5062 15.3537 10.7891 14.9849V11.9878H7.82422V13.019H9.43555V14.5498C9.14193 14.8649 8.64779 15.0225 7.95312 15.0225C7.31934 15.0225 6.8234 14.7897 6.46533 14.3242C6.11084 13.8551 5.93359 13.1981 5.93359 12.353V11.7515C5.94434 10.9028 6.10726 10.2601 6.42236 9.82324C6.74105 9.38281 7.20475 9.1626 7.81348 9.1626C8.75521 9.1626 9.30127 9.63346 9.45166 10.5752H10.7837C10.6799 9.76595 10.3719 9.14648 9.85986 8.7168C9.34782 8.28711 8.65674 8.07227 7.78662 8.07227C6.77327 8.07227 5.98372 8.40527 5.41797 9.07129C4.85221 9.73372 4.56934 10.6558 4.56934 11.8374V12.4497C4.58008 13.1873 4.72152 13.8337 4.99365 14.3887C5.26937 14.9401 5.65967 15.3644 6.16455 15.6616C6.66943 15.9588 7.24951 16.1074 7.90479 16.1074C8.54216 16.1074 9.11149 16.0125 9.61279 15.8228ZM12.3628 16H13.7163V8.17969H12.3628V16ZM16.7456 12.686H19.8931V11.6011H16.7456V9.27539H20.355V8.17969H15.3867V16H16.7456V12.686Z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
    </div>
  </body>,
  "container": .css-0 {
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;
}

<div>
    <button
      class="tweet-gif css-0"
      type="button"
    >
      <svg
        fill="#C4C4C4"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 2H21C21.5523 2 22 2.44772 22 3V21C22 21.5523 21.5523 22 21 22H3C2.44772 22 2 21.5523 2 21V3C2 2.44772 2.44772 2 3 2ZM0 3C0 1.34315 1.34315 0 3 0H21C22.6569 0 24 1.34315 24 3V21C24 22.6569 22.6569 24 21 24H3C1.34315 24 0 22.6569 0 21V3ZM9.61279 15.8228C10.1141 15.633 10.5062 15.3537 10.7891 14.9849V11.9878H7.82422V13.019H9.43555V14.5498C9.14193 14.8649 8.64779 15.0225 7.95312 15.0225C7.31934 15.0225 6.8234 14.7897 6.46533 14.3242C6.11084 13.8551 5.93359 13.1981 5.93359 12.353V11.7515C5.94434 10.9028 6.10726 10.2601 6.42236 9.82324C6.74105 9.38281 7.20475 9.1626 7.81348 9.1626C8.75521 9.1626 9.30127 9.63346 9.45166 10.5752H10.7837C10.6799 9.76595 10.3719 9.14648 9.85986 8.7168C9.34782 8.28711 8.65674 8.07227 7.78662 8.07227C6.77327 8.07227 5.98372 8.40527 5.41797 9.07129C4.85221 9.73372 4.56934 10.6558 4.56934 11.8374V12.4497C4.58008 13.1873 4.72152 13.8337 4.99365 14.3887C5.26937 14.9401 5.65967 15.3644 6.16455 15.6616C6.66943 15.9588 7.24951 16.1074 7.90479 16.1074C8.54216 16.1074 9.11149 16.0125 9.61279 15.8228ZM12.3628 16H13.7163V8.17969H12.3628V16ZM16.7456 12.686H19.8931V11.6011H16.7456V9.27539H20.355V8.17969H15.3867V16H16.7456V12.686Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
